### DISCUSSION TITLE
Evaluation of usage scenarios

### DISCUSSION THREAD

**Discussion by GeziP** [eProsima/Fast-DDS#3842] [Support]
Hi, I'm Lucas. We have a scenario where we want to use it for instrument control. There will be approximately four ARM Linux hosts connected over the network, each running dozens of services. We have the following requirements:

    1.Sequential Consistency
    2.Every message must be acknowledged, so should we use a request-reply model?
    3.Response time for some messages needs to be around 1ms. What would be the response time if shared memory (SHM) is used within the same machine? And what about TCP/UDP between different machines?

We're not sure if this is suitable for our needs.

**Comment by JLBuenoLopez** [comment]
Hi @GeziP 

Thanks for opening the ticket. I would try to redirect you to the corresponding sections of the documentation in order to answer your questions:

1. DDS protocol ensures that the samples are received in the same order as they have been written.
2. DDS is data-centric communication protocol so the request-reply model is not its optimal use. Nevertheless you can read more about how to set up a request-reply paradigm using Fast DDS in this [section of Fast DDS documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/request_reply/request_reply.html#use-case-request-reply). However, maybe it is enough if you use the [Reliability QoS Policy](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/standardQosPolicies.html#reliabilityqospolicy) configured as RELIABLE so every sample is acknowledged.
3. [This article in eProsima webpage](https://eprosima.com/index.php/resources-all/performance/eprosima-fast-dds-performance) includes some performance comparison using several configurations.

You might also consider contacting [eProsima support team](https://github.com/eProsima/Fast-DDS#getting-help) in order to discuss your questions in a meeting.

**Reply by GeziP** [reply]
Certainly, thank you for the answers. Additionally, I have a few more questions:

1. Is it necessary to meet the following conditions for sequence consistency?
   1. The value of `max_samples` must be higher or equal to the value of `max_samples_per_instance`.
   2. The value established for the `HistoryQosPolicy` `depth` must be lower or equal to the value stated for `max_samples_per_instance`?

2. Does enabling `RELIABLE_RELIABILITY_QOS` affect throughput? Is it no longer necessary to enable the reply-request mode after enabling it?

3. We have been using traditional RPC methods for instrument control. We are newcomers to the pub-sub model. If we transition to a data-centric approach, do you have any larger demos or examples that meet similar requirements?

**Comment by JLBuenoLopez** [comment]
1. Instances are another different concept within the DDS paradigm. You can find more information in [this section of Fast DDS documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/topic/instances.html). The History depth would have to be set to a proper value to ensure that no sample overwriting occurs. But that would depend on the behavior of the user application. eProsima offers [architecture studies](https://eprosima.com/index.php/services-all) where we help users tune their application up with the best QoS meeting their requirements and needs. You might be interested on contacting our support team for more info.
2. Throughput can be affected by the Reliability QoS policy depending on the behavior of the network. If the network is lossy, samples lost should have to be resent. There are possible workarounds but again, an architecture study will be needed to tune optimally your application.
3. I do not understand which requirements are you referring to. A Publisher-Subscriber model does not require to know who is publishing or who is receiving. It is a completely different paradigm from RPC. If you want to see how to use Fast DDS in a request-reply paradigm you have [this example in Fast DDS repository](https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/dds/RequestReplyExample) (the documentation link I provided above already pointed you to this example).

**Reply by GeziP** [reply]
Okay, thank you for your response. We are currently in the feasibility assessment phase, and if we need further support, we will get in touch.
